Sources/NIOPosix/PendingDatagramWritesManager.swift

```diff
@@ -515,14 +541,28 @@ final class PendingDatagramWritesManager: PendingWritesManager {
     ///
     /// - parameters:
     ///     - scalarWriteOperation: An operation that writes a single, contiguous array of bytes (usually `sendto`).
-    private func triggerScalarBufferWrite(scalarWriteOperation: (UnsafeRawBufferPointer, UnsafePointer<sockaddr>, socklen_t, AddressedEnvelope<ByteBuffer>.Metadata?) throws -> IOResult<Int>) rethrows -> OneWriteOperationResult {
+    private func triggerScalarBufferWrite(scalarWriteOperation: (UnsafeRawBufferPointer, UnsafePointer<sockaddr>?, socklen_t, AddressedEnvelope<ByteBuffer>.Metadata?) throws -> IOResult<Int>) rethrows -> OneWriteOperationResult {
         assert(self.state.isFlushPending && self.isOpen && !self.state.isEmpty,
                "illegal state for scalar datagram write operation: flushPending: \(self.state.isFlushPending), isOpen: \(self.isOpen), empty: \(self.state.isEmpty)")
         let pending = self.state.nextWrite!
         do {
-            let writeResult = try pending.address.withSockAddr { (addrPtr, addrSize) in
-                try pending.data.withUnsafeReadableBytes {
-                    try scalarWriteOperation($0, addrPtr, socklen_t(addrSize), pending.metadata)
+            let writeResult: IOResult<Int>
+            if self.state.isConnected {
```

Comment:
I think in this case we want to confirm that any write we're about to send has an address that matches the address we connected to.

Reply:
Oh, I was thinking that this would be an opportunity to get a small perf win. I realise there'll be a perf win within the call to `sendmsg`, but by not bothering with the `withSockAddr` here we avoid a copy. 

This comes down to a question about the layer of abstraction we want to provide here I suppose. Do we want to provide an API that checks you are doing the right thing or be more transparent?

I guess we are already abstracting `sendmsg` calls away because calling them with a `msg_name` will cause an error.

WDYT about checking that the address matches on the way _in_ to PDWM, in the call to `add` if it currently `isConnected`? This would make the actual writing out of data quicker?

Now we're talking about this, it's going to be even more complicated/costly in the vector write case to make this check for each of the message headers.

Do you think there's any milage in an API that, once you call `connect()` sends all datagrams to that address? This would be the most performant, but I wonder if it might mean we should be exposing different APIs for sending these messages (that don't take an `AddressedEnvelope`) and disallow the `AddressedEnvelope` calls when connected?

Reply:
Is a4ee0756d what you had in mind?

Reply:
@simonjbeaumont Yeah the performance win is valuable, but I think we need to detect misuse of the API as well.

If we wanted to get the performance win I think we get it by allowing sending naked byte buffers instead of addressedenvelopes once we connect. If users still send addressed envelopes then we still have to check. I'm fine with us doing the check on the way into PDWM, but if we do that we also have to validate anything that's buffered when the `connect` call comes in.

Reply:
Hm, the presence of buffering makes this an interesting problem since it's possible the connected state of the socket can change while things are buffered. That's the case even with what's in this PR, but becomes even more apparent if we start supporting `disconnectx` and re-`connect`.

Reply:
OK so I reverted the commit that checks the writes when flushing and replaced it with ae1b8d409 which fails any buffered writes when connecting or reconnecting the datagram channel and validates the destination address on the way in, when the channel is connected.

Also added some tests.